src/Api/Endpoints/Projects/Instruments.php

```diff
@@ -0,0 +1,165 @@
+ * @license  Loris license
+ * @link     https://github.com/aces/Loris
+ */
+class Instruments extends Endpoint implements \LORIS\Middleware\ETagCalculator
+{
+    /**
+     * A cache of the results of the endpoint, so that
+     * it doesn't need to be recalculated for the ETag and handler
+     */
+    protected $responseCache = array();
+
+    /**
+     * Return which methods are supported by this endpoint.
+     *
+     * @return array supported HTTP methods
+     */
+    protected function allowedMethods() : array
+    {
+        return ['GET'];
+    }
+
+    /**
+     * Versions of the LORIS API which are supported by this
+     * endpoint.
+     *
+     * @return array a list of supported API versions.
+     */
+    protected function supportedVersions() : array
+    {
+        return [
+                "v0.0.1",
+                "v0.0.2",
+                "v0.0.3-dev",
+               ];
+    }
+
+    /**
+     * Handles a request that starts with /projects/$projectname/candidates
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        // FIXME: Validate project based permissions.
+
+        $pathparts = $request->getAttribute('pathparts');
+        if (count($pathparts) == 1) {
+            return (new \LORIS\Http\Response())
+                ->withHeader("Content-Type", "application/json")
+                ->withBody(
+                    new \LORIS\Http\StringStream(
+                        json_encode(
+                            $this->_toArray($request)
+                        )
+                    )
+                );
+        }
+
+        // Delegate to project specific endpoint.
+        $endpoint = new Instrument();
+        array_shift($pathparts);
+        $request = $request->withAttribute('pathparts', $pathparts);
+
+        return $endpoint->process($request, $endpoint);
+    }
+
+    /**
+     * Create an array representation of this endpoint's reponse body
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return array
+     */
+    private function _toArray(ServerRequestInterface $request): array
+    {
+        $project     = $request->getAttribute('project');
+        $projectname = $project->getName();
+
+        if (!isset($this->responseCache[$projectname])) {
+
+            $meta = array('Project' => $projectname);
+
+            $ddeinstruments = \NDB_Factory::singleton()
+                ->config()
+                ->getSetting("DoubleDataEntryInstruments");
+
+            // FIXME: We should be able to instanciate an instrument and call
+            // ->isDDE() and ->getSubgroupName() instead of querying the
+            // persistence layer.
+            $result = \NDB_Factory::singleton()->database()->pselect(
```

Comment:
feels like this could use a provisioner too so that there's less work when we start adding project permissions, but this is less immediately applicable than the candidate one above..